src/osd/OSD.cc

```diff
@@ -6408,6 +6408,12 @@ void OSD::ms_fast_dispatch(Message *m)
     dout(10) << "ping from " << m->get_source() << dendl;
     m->put();
     return;
+  case MSG_MON_COMMAND:
+    handle_command(static_cast<MMonCommand*>(m));
+    return;
+  case MSG_COMMAND:
+    handle_command(static_cast<MCommand*>(m));
```

Comment:
This can immediately invoke con->send_message(). Embarrassingly, I've forgotten if we're allowed to do that inside of fast dispatch, but I don't think so?

Reply:
it's safe, we do it all over the place

Reply:
We definitely don't do it all over the place, and I'm quite sure we couldn't safely send messages out of fast_dispatch in the initial implementation, which was part of why we had all the elaborate tooling around scheduling map sharing to happen later on (```struct send_map_on_destruct``` put a ```C_SendMap``` in the work queue in Jewel, for instance). It *might* have become safe since then (especially with the Connection-based functions or AsyncMessenger), but those races were subtle and unpleasant so we'll need to either avoid doing so or work through the locking again to prove it.

The only place we've ever done so that I can find is an error condition we wouldn't have tested: we return ENXIO on misdirected ops in-line. That in fact was something I kept around, but I think I just missed it.

(Really kicking myself for not documenting what was allowed here, since I thought we'd done such a good job dispatching it. But I think I just assumed sending messages fell under "long-held locks", and there were a bunch of cyclical deadlocks that involved a PG lock held while sending off messages while we were trying to dispatch a message to them...)

Reply:
Hmm, yeah I remember it being weird, but I also can't remember why.  The first case I remember reviewing or doing this is the fast dispatch for osd heartbeats which went in a while back.. see handle_osd_ping(), which will also send replies.

Stepping back, I don't see any reason why it *shouldn't* be possible.  If there are remaining reasons (like loopback causing deadlock/recursion) then IMO we should fix that instead!

Reply:
The global SimpleMessenger::lock is the obvious reason it would have caused trouble. The Connection-based interfaces are a lot safer, but still not guaranteed since they have to grab a bunch of extra locks if the pipe failed and needs to reconnect.

Handling reconnect means that guaranteeing a message dispatch is non-blocking is basically impossible unless we switch that model around completely; even if that doesn't cause a deadlock it's still distinctly undesirable — and it makes hard-to-identify deadlocks a lot more likely.

I'm not really sure I see a path that lets us generically guarantee that sending messages from fast_dispatch is safe. We could probably do it (by which I mean, it probably already works) if we restrict it to lossy connections. But I hate making use of interfaces which are that subtle and easy to break.

Fast dispatch works reasonably well because we made a simple interface: you're not allowed to do anything that might block. No matter how good we are, we can't send messages without doing something that might block. Therefore, we can't send messages in fast dispatch. It's the locking boundary that makes an interface like this between the OSD and Messenger possible.

Reply:
I'm not really following. What does reconnect or lossy/not lossy have to do with it?  send_message() just needs to drop the message into a queue to be sent out (and that's roughly what it does for both simple and async).  The messenger implementation just needs to make sure the fast dispatching thread isn't holding locks that would deadlock against send_message().  Which apparently they don't, since lockdep doesn't explode with either async or simple, despite the heartbeat_dispatch() sends that have been there for a while now.

Reply:
Well, in SimpleMessenger if there's been a failure, we have to set up a new Pipe in-line because otherwise we don't *have* a queue to drop the Message in. I assume AsyncMessenger is similar.
We *could* try and do more exciting things involving another thread that polls for Messages that need to go out on non-existent Pipes...

As I said, this doesn't mean it's going to fail very often, and I think the heartbeat dispatch probably slots in to the "happens to work" category, since it uses lossy connections and doesn't have any other threads talking to that messenger. It's just not a very safe or useful interface in general.

Reply:
I will note that in a seastar world this is much easier because if we have to do something blocking like establishing a new connection, it's easy to queue it to happen later. We just don't have that infrastructure right now.

Reply:
And oh yeah, pipe replacement requires us to stop the previous Pipe's fast dispatch. There we go, deadlock.

Since we already have an easy model for queuing messages to go out not in the fast dispatch path, that seems a better idea right now than a lot of work that will destabilize our existing code and interfaces.

Reply:
Oh bother.. I remember those deadlocks now.  Yeesh.

Okay... so I think the primary value here is that it is funneling everything through the queue, and the message handlers are just generating queue events.  I'll move these back to slow dispatch but leave the rest the same.  I don't think that will make things any easier or harder in the eventual seastar transition.  The hard part there is going to be handle_osd_map and the various bits of OSDService that have their own special locks (and aren't sharded).

Reply:
That works too, and yes, I don't expect it to have any real impact on our future work!

Reply:
...and yeah MSG_COMMAND looks like the only one, so no biggie!